numba/untyped_passes.py

```diff
@@ -415,6 +417,134 @@ def run_pass(self, state):
         return False
 
 
+@register_pass(mutates_CFG=True, analysis_only=False)
+class CanonicalizeLoopExit(FunctionPass):
+    """A pass to canonicalize loop exit by splitting it from function exit.
+    """
+    _name = "canonicalize_loop_exit"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    def run_pass(self, state):
+        fir = state.func_ir
+        cfg = compute_cfg_from_blocks(fir.blocks)
+        status = False
+        for loop in cfg.loops().values():
+            for exit_label in loop.exits:
+                if exit_label in cfg.exit_points():
+                    self._split_exit_block(fir, cfg, exit_label)
+                    status = True
+
+        fir._reset_analysis_variables()
+
+        vlt = postproc.VariableLifetime(fir.blocks)
+        fir.variable_lifetime = vlt
+        return status
+
+    def _split_exit_block(self, fir, cfg, exit_label):
+        curblock = fir.blocks[exit_label]
+        newlabel = exit_label + 1   # XXX
+        newlabel = find_max_label(fir.blocks) + 1
+        fir.blocks[newlabel] = curblock
+        newblock = ir.Block(scope=curblock.scope, loc=curblock.loc)
+        newblock.append(ir.Jump(newlabel, loc=curblock.loc))
+        fir.blocks[exit_label] = newblock
+        # Rename all labels
+        fir.blocks = rename_labels(fir.blocks)
+
+
+@register_pass(mutates_CFG=True, analysis_only=False)
+class CanonicalizeLoopEntry(FunctionPass):
+    """A pass to canonicalize loop header by splitting it from function entry.
+
+    This is needed for loop-lifting; esp in py3.8
+    """
+    _name = "canonicalize_loop_entry"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    def run_pass(self, state):
+        fir = state.func_ir
+        cfg = compute_cfg_from_blocks(fir.blocks)
+        status = False
+        for loop in cfg.loops().values():
+            if len(loop.entries) == 1:
```

Comment:
Should this have an `else: assert 0 'unreachable'` or similar?

Reply:
This is intended to only work in the cases that it knows how, and ignore the other cases. I don't think it should raise unreachable here.  But, would be good to know if we can rule out multi-entry loops in controlflow.py

Reply:
Agree. I also don't recall seeing a multi-entry loop in the wild but imagine they could occur synthetically.